Add Physics base class and a heat conduction example

GiudGiud opened the pull request:
Open for comments.
Once the dependent PRs are in, this could actually be merged
refs https://github.com/idaholab/moose/issues/25642
see original PR https://github.com/idaholab/moose/pull/25704

I dont think the syntax has been decided yet, so I m open to a CCB meeting on this

Currently the Physics base class only implements:
- some basic parameter checking
- an insertion point for relationship managers at the equation level (could be removed for now, will come in for NS)
- an insertion point for restarting from Exodus (could be removed for now, will come in for NS)

GiudGiud committed:
Add a PhysicsName type that will be used to pass Physics to other objects
refs #25642

GiudGiud committed:
Add a base class for defining Physics
refs #25642

GiudGiud committed:
Add some base routines for restarting variables in a Physics
refs #25642

GiudGiud committed:
Add documentation to Physics
refs #25642

GiudGiud committed:
Add a system number to the Physics
refs #25642

GiudGiud committed:
Add 1st utility to check parameters
refs #25642

GiudGiud committed:
Add another set of utilities to check parameters:
- length of vector params
- defined or not
refs #25642

GiudGiud committed:
Add a base class for heat conduction equation
refs #25642

GiudGiud committed:
Add a derived class for CGFE heat conduction
refs #25642

GiudGiud committed:
Add documentation for the finite element heat conduction physics
refs #25642

GiudGiud committed:
Add a test for the heat conduction physics
refs #25642

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1865010/) on 0ac3227 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25977/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1884642/) on 9a218a6 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ed6922</th>
      <th colspan=4>#25977 9a218a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.74%</td>
      <td>85.62%</td>
      <td>-0.13%</td>
      <td>1.23%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>93035</td>
      <td>93037</td>
      <td>+2</td>
      <td>2</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15469</td>
      <td>15629</td>
      <td>+160</td>
      <td>160</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25977/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25977/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>ed6922</th>
      <th colspan=4>#25977 9a218a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>88.42%</td>
      <td>88.24%</td>
      <td>-0.19%</td>
      <td>79.79%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3720</td>
      <td>3795</td>
      <td>+75</td>
      <td>75</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>487</td>
      <td>506</td>
      <td>+19</td>
      <td>19</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25977/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25977/coverage/heat_transfer)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25977/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 1.23% is less than the suggested 90.0%
- `heat_transfer` new line coverage rate 79.79% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud committed:
Duplicate physics source page as a syntax page
refs #25642

GiudGiud commented:
Should probably move this forward. It doesnt have to, we could also wait until I dig deep into Component as Actions and assess if this is going to change the deal for Physics. I dont expect it to.

lindsayad reviewed (commented):
want to add a framework test?

lindsayad commented:
The thing that came to my mind is `DiffusionPhysics` or `PoissonPhysics`. @nmnobre just recently created a `coupled_elestrostatics.i` framework level test in his Raviart-Thomas PR.... It could be a good demonstration of `Physics` to make both CFEM and Raviart-Thomas implementations of the Poisson equation.... Although I can't remember if you wanted `Physics` to be useful for the purpose of allowing different algebraic discretizations of the same physics. Also you probably don't want to balloon this PR too much, so maybe just a CFEM implementation of Poisson would be good to start, and then we could add RT later.

I thought we had an issue where we had a pretty large community (I thought it involved @dschwen at least) discussion about the Physics class but maybe i've lost my marbles. #25642 doesn't seem like it.

GiudGiud commented:
I'm happy to implement that.

GiudGiud commented:
A lot of the discussion is on the original Physics PR https://github.com/idaholab/moose/pull/25704

> wanted Physics to be useful for the purpose of allowing different algebraic discretizations of the same physics

I do want that. I dont think it s feasible to do a `[Discretization]` block with our current ways of parsing the input file without being very convoluted

lindsayad commented:
> A lot of the discussion is on the original Physics PR #25704

Ah there it is

GiudGiud committed:
Add a base class for solving the diffusion equation

GiudGiud committed:
Add a derived physics class to solve diffusion with CG

GiudGiud committed:
Add a derived physics class to solve diffusion with FV

GiudGiud committed:
Add tests for diffusion example physics

GiudGiud committed:
Allow for a parameter rename when transferring a parameter

GiudGiud committed:
Set some default preconditioning options for solving diffusion

GiudGiud committed:
Add documentation to example diffusion physics, refs #25642

GiudGiud commented:
Added a FV and CG example in the framework. might not pass tests yet

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit e4812f5b9245831473dd269d4bfd5e8485f536b4
+Subproject commit 47379e241204355cd5701027bf85f222718162ec
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1937670/) on f2ecfcc wanted to post the following:

The following commit(s) contain fixups:

- 52cdfbe734: fixup! Add documentation to example diffusion physics, refs #25642

GiudGiud committed:
Fix documentation:
It has to be: / name of block / syntax used / name of action. So here /Physics/DiffusionCG/DiffusionCG

GiudGiud committed:
Use preconditioning by default for diffusion

GiudGiud commented on framework/src/physics/DiffusionCG.C:
```diff
@@ -0,0 +1,231 @@
+registerMooseAction("MooseApp", DiffusionCG, "add_kernel");
+registerMooseAction("MooseApp", DiffusionCG, "add_bc");
+registerMooseAction("MooseApp", DiffusionCG, "add_variable");
+registerMooseAction("MooseApp", DiffusionCG, "add_preconditioning");
```
it would be good to inherit this one from the base class. Unfortunately this is not how it works.
We cannot just register the task to `[Physics]` as the parameters for the preconditioning are nested under `Physics/DiffusionCG/some_name`. 
It does not seem to me that I can register to DiffusionPhysicsBase either. 

Maybe the way forward is to add the task dynamically at construction?

lindsayad replied:
I think there are cases where an action inherits from another where you don't want to automatically inherit its associated tasks. I believe the `add_bounds` task is one I reviewed of yours today where you were deliberately trying to disambiguate bounds from kernel/auxkernel but you inherit the `AddBoundsAction` from `AddKernelAction`.

You could define a macro in the base class
```c++
#define registerDiffusionPhysicsBaseTasks(derived_name) \
  registerMooseAction("MooseApp", derived_name, "add_preconditioning")
```

GiudGiud replied:
That works. I ll add that

lindsayad replied:
I was trying to understand this afternoon why the coverage is relatively low despite your framework `Physics` class and testing. It is tied to the task registration. If a developer implements one of your `virtual` methods without registering the task then they will likely be surprised when their objects don't get added. So maybe we do want tasks registered in a base action to, by default, propagate to the derived class. And then we could offer a way to de-register an inherited task if desired. What do you think? If you're in agreement, do you want to do that here in this PR or we can create an issue?

lindsayad replied:
So essentially I think all the tasks that you check in `PhysicsBase::act` should be associated tasks for derived classes. This feels right given your `virtual` interfaces; e.g. a developer who implements any of your `virtual` methods will get the behavior they might expect without having to remember to register a task or call a macro from the base class. This would also have a side effect of making the coverage in `PhysicsBase::act` 100%

GiudGiud replied:
> This would also have a side effect of making the coverage in PhysicsBase::act 100%

actually since I would de-register DiffusionCG and DiffusionFV from all the tasks they do not use, and there are no other Physics in the framework, I would not get any better coverage from this.

GiudGiud replied:
I do agree that this is how the `virtual` interfaces should work.

lindsayad replied:
> actually since I would de-register DiffusionCG and DiffusionFV from all the tasks they do not use

Why would you do this? You have empty implementations for the virtual methods in the base class so why should a user need to bother de-registering tasks? It seems like a user should only have to think about implementing the virtual methods and that's that

GiudGiud replied:
looks better in the debug log. You dont see plenty of tasks being triggered doing nothing

lindsayad replied:
Well I still think there should be registration of tasks in `PhysicsBase` and that they should be inherited. The macro is a workaround, but I think it's a poor one. We need to make sure that tasks like your `init_physics` task should run for all physics. Right now someone who calls your `dimension()` API and relies on the result will get undefined behavior because your `_dim` member is uninitialized

GiudGiud replied:
I agree with that. 
Just letting you know I plan to deregister as well

lindsayad replied:
fair enough

GiudGiud replied:
Thinking more about this. 
I m thinking register everything remotely available / needed (all the tasks for the add...) in PhysicsBase
then if the relevant routine is not implemented, deregister automatically.

this would require making act() final too.
and I dont think "check if a routine is implemented and not just {} " is possible in C++ from my googling

back to thinking

GiudGiud replied:
From the registerMooseAction macro we can see there's a lot of meta data that we are not going to be able to get by simply adding a call in the parent class that would work for all the derived classes. 
```
      Registry::addAction<classname>({app, #classname, "", task, __FILE__, __LINE__, "", ""})
```
classnames, file and line are not something the parent class knows. Even the app may not be the same app.
Then file and line are only for debugging so that s something I can overlook for now.

That data is later processed (into the action warehouse I think) into a list of tasks per action.
I think if we did something, we d work in the action warehouse directly instead of the registry because of the missing data. Still seems tricky to have calls in the base action that register tasks for derived actions.

So Instead, let's allow ourselves to have the calls registering the tasks in the derived actions. A few options:
- listing every task every time
Not super clean. Especially since act() is no longer defined in the derived class.

- the macro you suggested
The macro does make it a one-liner, we are still adding additional tasks on top of it

- looping over base classes of the physics to register the tasks of the base classes
not possible in C++ based on my googling. You d have to list them explicitly

I think I want to have option 2 + the following changes
- make the PhysicsBase addXXX return a warning if they are called. It means too many tasks got registered
- keep track of the required tasks for a physics class, and have a check that all the required tasks are registered for the derived classes.

GiudGiud replied:
@loganharbour @dschwen 

If there is a better way to 'inherit' these tasks from the parent physics class.
registerMooseAction is necessarily called from the derived class. 
looping over all actions and dynamic_casting is somewhat of an option but not my favorite as it would mean registering tasks to the virtual parent classes (and checking all the derived ones against each of their parent)

GiudGiud committed:
More catering to that unit test that enforces MOOSE does not add new systems
refs #26033

GiudGiud committed:
Change moose server unit test so it no longer enforces the list of systems refs #26033

GiudGiud commented on framework/doc/content/source/physics/DiffusionCG.md:
```diff
@@ -0,0 +1,49 @@
+# DiffusionCG
+
+!syntax description /Physics/DiffusionCG/DiffusionCG
```
this is one place where I deviate from an Action because the name of the syntax is the same as the name of the object. Your typical action here would be `/Kernels/AddKernelAction` for example.

Here we have a Physics which does work like an action (and inherits from Action), but the doco is actually meant to be looked at (unlike let's AddBCAction, which is not very interesting in itself and forwards to the BC page) so I want it documented under Physics/its own name. 

So the traditional MOOSE way here would be:
- syntax documentation for Physics, does not show up in source index under physics (unless we symlink it there?)
- instead of calling the Physics object DiffusionCG, it'd be called AddDiffusionCGAction. But this is not great:

- a DiffusionCG is not a MooseObject, unlike each XXX that the AddXXXAction adds usually. So all the code for the DiffusionCG is in the "Action" / Physics, and not in another class.

lindsayad replied:
I can't tell if you're looking for feedback here?

GiudGiud replied:
Always happy to get feedback. I dont know how to do better but maybe you or @dschwen will know a better way

lindsayad replied:
I can't think of anything off the top of my head. Hopefully over time we can come up with something better. I'm not too worried about it at the moment

GiudGiud commented:
Timeouts in debug unrelated

GiudGiud committed:
Use a macro to propagate base classes tasks to derived ones

GiudGiud committed:
Keep track of required tasks in the Physics objects
Fix two typos

GiudGiud committed:
Add preconditioning to heat conduction physics too

lindsayad commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -0,0 +1,248 @@
+  /// Check that the two vector parameters are of the same length
+  template <typename T, typename S>
+  void checkVectorParamsSameLength(const std::string & param1, const std::string & param2) const;
+  /// Check that there is no overlap between the two vector parameters
+  template <typename T>
+  void checkVectorParamsNoOverlap(const std::vector<std::string> & param_vec) const;
```
It looks like the doxygen should be updated to say this is checking that all elements of the vector are unique

GiudGiud replied:
I think it s a vector of names of vector parameters

GiudGiud replied:
I ll check

GiudGiud replied:
done

lindsayad commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -0,0 +1,248 @@
+
+  /// Dimension of the physics, which we expect for now to be the dimension of the mesh
+  /// NOTE: this is not known at construction time, only after initializePhysics which is a huge bummer
+  unsigned int _dim;
```
```suggestion
  unsigned int _dim = libMesh::invalid_uint;
```
or do something like this in the constructor so that your assertion in `dimension()` works

GiudGiud replied:
yeah this was probably done in the original PR. Forgot to move it

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+{
+  InputParameters params = HeatConductionPhysics::validParams();
+  params.addClassDescription("Creates the heat conduction equation discretized with CG");
+
```
```suggestion
```

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+    InputParameters params = getFactory().getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _temperature_name;
+    params.applyParameter(parameters(), "thermal_conductivity");
+    getProblem().addKernel(kernel_type, name() + "_" + _temperature_name + "_conduction", params);
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+    InputParameters params = getFactory().getValidParams(kernel_type);
+    params.set<NonlinearVariableName>("variable") = _temperature_name;
+    params.set<std::vector<VariableName>>("v") = {getParam<VariableName>("heat_source_var")};
+    getProblem().addKernel(kernel_type, name() + "_" + _temperature_name + "_source", params);
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+    params.set<NonlinearVariableName>("variable") = _temperature_name;
+    params.applyParameter(parameters(), "density_name");
+    params.applyParameter(parameters(), "specific_heat");
+    getProblem().addKernel(kernel_type, name() + "_" + _temperature_name + "_time", params);
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+      params.set<std::vector<BoundaryName>>("boundary") = heat_flux_boundaries;
+      params.set<MooseFunctorName>("functor") = boundary_heat_fluxes[0];
+      getProblem().addBoundaryCondition(
+          bc_type, name() + "_" + _temperature_name + "_heat_flux_bc_all", params);
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+        params.set<std::vector<BoundaryName>>("boundary") = {heat_flux_boundaries[i]};
+        params.set<MooseFunctorName>("functor") = boundary_heat_fluxes[i];
+        getProblem().addBoundaryCondition(bc_type,
+                                          name() + "_" + _temperature_name + "_heat_flux_bc_" +
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+      params.set<std::vector<BoundaryName>>("boundary") = temperature_boundaries;
+      params.set<MooseFunctorName>("functor") = boundary_temperatures[0];
+      getProblem().addBoundaryCondition(
+          bc_type, name() + "_" + _temperature_name + "_dirichlet_bc_all", params);
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionFE.C:
```diff
@@ -0,0 +1,142 @@
+        params.set<std::vector<BoundaryName>>("boundary") = {temperature_boundaries[i]};
+        params.set<MooseFunctorName>("functor") = boundary_temperatures[i];
+        getProblem().addBoundaryCondition(bc_type,
+                                          name() + "_" + _temperature_name + "_dirichlet_bc_" +
```
prefix

lindsayad commented on modules/heat_transfer/src/physics/HeatConductionPhysics.C:
```diff
@@ -0,0 +1,85 @@
+  if (_preconditioning == "default")
+  {
+    // We only pass petsc options as that's all that's needed to set up the preconditioner
+    Moose::PetscSupport::PetscOptions & po = _problem->getPetscOptions();
```
```suggestion
    auto & po = _problem->getPetscOptions();
```

lindsayad commented on test/tests/physics/diffusion_cg.i:
```diff
@@ -0,0 +1,89 @@
+    type = BreakBoundaryOnSubdomainGenerator
+    input = cmg
+  []
+  allow_renumbering = false
```
why?

GiudGiud replied:
NodalValue sampler cant be trusted with distributed mesh otherwise

lindsayad commented on test/tests/physics/diffusion_cg.i:
```diff
@@ -0,0 +1,89 @@
+[]
+
+[Outputs]
+  exodus = true
+  csv = true
```
```suggestion
  csv = true
```

lindsayad commented on test/tests/physics/diffusion_fv.i:
```diff
@@ -0,0 +1,91 @@
+[]
+
+[Outputs]
+  exodus = true
```
```suggestion
```

lindsayad commented on test/tests/physics/tests:
```diff
@@ -0,0 +1,25 @@
+    input = 'diffusion_cg.i'
+    cli_args = 'Physics/DiffusionCG/diff/automatic_differentiation=false Physics/DiffusionCG/diff/source_functor=value_pp'
+    csvdiff = 'diffusion_cg_out_sample_0010.csv'
+    requirement = 'The system shall be able to solve a source-less diffusion equation with a continuous Galerkin discretization.'
```
```suggestion
    requirement = 'The system shall be able to solve a source-less diffusion equation with a continuous Galerkin discretization with a shorthand syntax.'
```

lindsayad commented on test/tests/physics/tests:
```diff
@@ -0,0 +1,25 @@
+    type = CSVDiff
+    input = 'diffusion_cg.i'
+    csvdiff = 'diffusion_cg_out_sample_0010.csv'
+    requirement = 'The system shall be able to solve a source-less diffusion equation with a continuous Galerkin discretization with automatic differentiation.'
```
```suggestion
    requirement = 'The system shall be able to solve a source-less diffusion equation with a continuous Galerkin discretization with automatic differentiation with a shorthand syntax.'
```

lindsayad commented on unit/src/MooseServerTest.C:
```diff
@@ -1052,7 +1052,8 @@ TEST_F(MooseServerTest, CompletionDocumentRootLevel)
   EXPECT_EQ(request_id, response_id);
 
-  EXPECT_EQ(48u, completions_array.size());
+  // If the array grows with new syntax, let it grow
+  EXPECT_TRUE(completions_array.size() > 48);
```
lol

lindsayad reviewed (changes requested):
I like it

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -0,0 +1,248 @@
+  /// NOTE: This does not register the task, you still need to call registerMooseAction
+  void addRequiredPhysicsTask(const std::string & task) { _required_tasks.insert(task); }
+
+  /// System number for the systems owning the variables
+  const unsigned int _sys_number;
```
currently only one. We ll change that when we add NS

lindsayad replied:
I might change the doxygen string then. "System number for the systems" is kind of funny

GiudGiud committed:
Address Alex review

- use physics prefix instead of name for physics-created objects
- be more careful with uninitialized members
- more const-ness
- properly implement additional ghosting requests for FV diffusion

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -0,0 +1,254 @@
+{
+  mooseDoOnce(checkRequiredTasks());
+
+  if (_current_task == "init_physics")
```
@loganharbour 

this is the first big change between physics and actions.
We'd expect Physics developers to implement the relevant routines and register the tasks, instead of re-writing some version of this act() routine

this act() looks a lot like other physics-defining actions (TM, NS)

GiudGiud commented on framework/src/physics/DiffusionPhysicsBase.C:
```diff
@@ -0,0 +1,85 @@
+  MooseEnum pc_options("default none", "default");
+  params.addParam<MooseEnum>(
+      "preconditioning", pc_options, "Which preconditioning to use for this Physics");
+
```
@loganharbour 

this is the second thing. I would have devs define parameters and anything shared in the relevant PhysicsBase class, and derived classes implement the actual discretization using objects

GiudGiud replied:
It's not something that is impossible in actions, it just has not been done much (at least in NS)

GiudGiud commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -0,0 +1,248 @@
+
+  /// Get the factory for this physics
+  /// The factory lets you get the parameters for objects
+  Factory & getFactory() { return _factory; }
+  Factory & getFactory() const { return _factory; }
```
done

GiudGiud replied:
actually the const in Factory works, but if we need to modify parameters I m not sure I can get rid of that non-const getter

GiudGiud commented on framework/src/physics/DiffusionFV.C:
```diff
@@ -0,0 +1,178 @@
+                                                : "1";
+    getProblem().addFVKernel(kernel_type, name() + "_" + _var_name + "_diffusion", params);
+    // This should not be necessary but it is.
+    Action::addRelationshipManagers(Moose::RelationshipManagerType::COUPLING, params);
```
this was done in the longer PR I think. I ll move that change over

YaqiWang commented on modules/heat_transfer/src/physics/HeatConductionPhysics.C:
```diff
@@ -0,0 +1,85 @@
+  InputParameters params = PhysicsBase::validParams();
+  params.addClassDescription("Add the heat conduction physics");
+
+  params.addParam<VariableName>("temperature_name", "T", "Variable name for the temperature");
```
This will be better if we can have a `block` parameter with which the heat conduction physics can be block restricted. Checks will be need to make sure all boundary sides of the subdomain are covered with boundary conditions. We had this in Griffin for neutronics so that we can do multi-scheme neutron transport calculations.

GiudGiud replied:
the block parameter exists, it's on the PhysicsBase class

I think those checks are valuable. They should have contributed to MOOSE. My vision for this would be a `[Diagnostics]` system that centralizes all these checks

GiudGiud replied:
the Physics would pull in the diagnostics that are standard for each pde

YaqiWang commented on modules/heat_transfer/test/tests/physics/test_fe.i:
```diff
@@ -0,0 +1,50 @@
+
+[Physics]
+  [HeatConductionFE]
+    [h1]
```
Do we need this `[h1]` level?

In Griffin, we have a syntax like
```
[TransportSystems]
  # equation type, bc, source, etc.
  [sn]
    block = 0
    scheme = DFEM-SN
    # angular quadrature, FE family, order, etc.
  []
  # multiple sub-blocks are allowed with proper block restriction
[]
```
`[TransportSystems]` is very like `[HeatConduction]` but we use a meta action to add individual actions based on the `scheme` parameter.

GiudGiud replied:
you do if you want to have multiple instances and want to vary those
```
  # equation type, bc, source, etc.
```
parameters

the nice thing is that you do have a 'discretization' block effectively. 

I think we could discuss whether it s better to have the discretization that way. it s a lot of nesting

```
[Physics]
  [Transport]
     [region_1_problem]
        block = 0
         # equation type, bc, source, etc.
         [FV]
            scheme = DFEM-SN
            # angular quadrature, FE family, order, etc.
          []
      []
    []
  []
  # multiple sub-blocks are allowed with proper block restriction
[]

YaqiWang replied:
`equation type, bc, source, etc` are defining the problem and shared by all subdomains. Each subdomain has a particular discretization scheme. This is the reason why we designed `[TransportSystems]` syntax that way. The thermal conduction, we could potentially have CG+DG multi-scheme. We may have only part of the domain have heat conduction.

YaqiWang replied:
I think it is fine to merge the equation or physics description with discretization as you did. We just do not need this extra input level `[h1]` here I think. It will be a little difficult to do multi-scheme but that may not be important.

GiudGiud replied:
> radiation type (neutron/photon/etc), equation type (eigenvalue/transient), bc, source, etc are defining the problem and shared by all subdomains.

that s not the case for other Physics we have. 

> . The thermal conduction, we could potentially have CG+DG multi-scheme

we can already do that with the current syntax. 

We need the nested block for block restriction

joshuahansel commented:
Overall design looks good to me. I wish the syntax could be
```
[Physics]
  [diff]
    type = DiffusionFE
```
instead of
```
[Physics]
  [DiffusionFE]
    [diff]
```
but I believe we've decided that Action is the right approach, which I think constrains the syntax to as you have.

Let me know when this is ready for a thorough review. I'll have a lot of requests for documentation, but I'd probably just have us fix anything wrong and then I can do a follow-up PR to improve the documentation to my satisfaction, and you can review that.

joshuahansel commented:
Oh, one thing I forgot: do you have a working example (not necessary to be in this PR) with the integration of this design with Components?

GiudGiud commented:
> Oh, one thing I forgot: do you have a working example (not necessary to be in this PR) with the integration of this design with Components?

other PR https://github.com/idaholab/moose/pull/25704

GiudGiud commented:
> but I believe we've decided that Action is the right approach, which I think constrains the syntax to as you have.

yes. And having `type=` forward to an action is difficult. 
Especially if we want:
```
  type = FV
  [Energy]
  ```
to pick an EnergyFV action, this is some multi-line parsing which we do not have

GiudGiud commented:
> Let me know when this is ready for a thorough review. I'll have a lot of requests for documentation, but I'd probably just have us fix anything wrong and then I can do a follow-up PR to improve the documentation to my satisfaction, and you can review that.

Alex has been thoroughly reviewing this. I dont think there is anything left wrong! He did catch a lot of wrong things

GiudGiud committed:
Update to new syntax based on discussions with Daniel, Kyle, Peter and Stefano

joshuahansel commented:
> I dont think there is anything left wrong! He did catch a lot of wrong things

Oh don't worry, there's wrong stuff still. But just doc/typos so far 😄

joshuahansel commented on framework/doc/content/source/physics/DiffusionCG.md:
```diff
@@ -0,0 +1,49 @@
+
+The volumetric discretization of the time derivative uses the [TimeDerivative.md] kernel.
+The diffusion term $\nabla \cdot D \nabla u(\vec{x})$ is integrated by parts and represented using:
+- a [Diffusion.md] kernel if no diffusion coefficient is passed
```
You need a blank line before each of these lists. Do this in all of your .md files.

joshuahansel commented on modules/heat_transfer/doc/content/source/physics/HeatConductionFE.md:
```diff
@@ -0,0 +1,19 @@
+# Heat Conduction Finite Element
```
I'd just keep it as "HeatConductionFE" to match the class name

GiudGiud replied:
Syntax is HeatConduction/FiniteElement so it's more searchable that way?
I think if you are coming from a moose input to the documentation this way is nice
If you are coming from the source code and wanting to look at the doco, HeatConductionFE is easier

GiudGiud replied:
I guess for Diffusion I did not do it. As you want.

GiudGiud replied:
Yeah it s the 'source' index so we ought to be consistent with the source

the syntax page can be consistent with the syntax.

joshuahansel commented on test/tests/physics/diffusion_fv.i:
```diff
@@ -0,0 +1,92 @@
+
+[Physics]
+  [Diffusion]
+    [FiniteVolume]
```
Is there a good reason to have a nested block for the spatial discretization, instead of doing
```
[Physics]
  [DiffusionFV]
```
To me, what you have makes more sense only if you plan to have multiple discretizations in the same input file, which seems unlikely.

GiudGiud replied:
At this point it s personal preferrence. 
```
[Physics]
  [DiffusionFV]
```
```
[Physics]
  [Diffusion/FV]
```
```
[Physics]
  [Diffusion]
    [FiniteVolume]
```

I think I got used to having `FV` but actually I think Daniel is right that spelling things out is better

GiudGiud replied:
this
```
[Physics]
  [Diffusion]
    [FiniteVolume].
```
could be a prelude to allowing this
```
[Physics]
  [Diffusion]
    [Default]
      [fuel]
```
or simply this
```
[Physics]
  [Diffusion]
    [fuel]
```

joshuahansel replied:
```
[Physics]
  [Diffusion]
    [FiniteVolume]
```
is fine. Maybe it's better for users. For devs, maybe it's just a little confusing sometimes to pair it with the class registered to that syntax.

lindsayad replied:
Although rarish, I could definitely imagine people mixing discretizations in a single input. For instance, in an FSI simulation, I may simulate the fluid with an FV discretization and so want heat transfer physics with FV in that subdomain, and then the solid with a CFEM discretization and so want a heat transfer physics with FE in that subdomain

GiudGiud committed:
Address Josh review
- fix documentation

Apply suggestions from code review

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

joshuahansel commented on framework/src/physics/DiffusionFV.C:
```diff
@@ -0,0 +1,194 @@
+{
+  InputParameters params = DiffusionPhysicsBase::validParams();
+  params.addClassDescription(
+      "Add the heat conduction physics discretized with cell-centered finite volume");
```
```suggestion
      "Add diffusion physics discretized with cell-centered finite volume");
```

joshuahansel commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -0,0 +1,254 @@
+  else
+    _dim = _mesh->dimension();
+
+  mooseAssert(_mesh, "We should have a problem to check if it is transient");
```
Probably wanted `_problem`, but `getProblem()` does this anyway

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2024467/) on 42df197 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25977/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25977/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c581c9a5ca5dd593c14f3e13409e066da16cfe61`

GiudGiud commented on framework/src/physics/DiffusionFV.C:
```diff
@@ -0,0 +1,194 @@
+  params.addClassDescription(
+      "Add the heat conduction physics discretized with cell-centered finite volume");
+  // No kernel implemented in the framework for a material property diffusivity
+  params.suppressParameter<MaterialPropertyName>("diffusivity_matprop");
```
Suppressing is fine

I like having all the properties defined in the base class.

GiudGiud commented:
I ll do one for all of my January work soon

lindsayad commented on framework/include/physics/PhysicsBase.h:
```diff
@@ -0,0 +1,248 @@
+  /// If you need more than these:
+  /// - register your action to the new task using
+  ///   registerMooseAction("AppName", ActionClass, "task_name");
+  /// - override act and add your additional work there
```
```suggestion
```

GiudGiud replied:
what's the plan if we need a derived Physics to use additional tasks? For example if we are dealing with an app that has systems that dont exist in moose/framework
I m thinking the 'final' will have to go (either one day or now)

GiudGiud replied:
While it is nice to pre-list the tasks, and safer because there's less chances for a mismatch (ignored tasks for which we implemented work is the main risk), we've never had this for actions. I could add some sort of boolean that checks that the main "act" of PhysicsBase always gets called. But that's unconventional?

GiudGiud replied:
actually I have a solution for this

GiudGiud replied:
added

GiudGiud committed:
Add a new insertion point for physics adding new tasks

GiudGiud committed:
Add proper syntax pages for everything but the last level syntax

GiudGiud commented:
noice thanks a lot for the reviews. Merging so it makes the newsletters

now for the tough part. NS, TM and THM
